src/coord/src/coord.rs

```diff
@@ -3067,15 +3067,30 @@ impl<S: Append + 'static> Coordinator<S> {
             })
             .transpose()?;
 
-        if compute_instance.replicas_by_id.is_empty() {
+        let source_ids = source.depends_on();
+
+        let num_replicas = compute_instance.replicas_by_id.len();
+        if num_replicas == 0 {
             return Err(CoordError::NoClusterReplicasAvailable(
                 compute_instance.name.clone(),
             ));
         }
 
-        let compute_instance = compute_instance.id;
+        // Reading from log sources on replicated compute instances is only
+        // allowed if a target replica is selected. Otherwise, we have no way
+        // of knowing which replica we read the introspection data from.
+        if num_replicas > 1 && target_replica.is_none() {
+            let log_names = source_ids
+                .iter()
+                .flat_map(|id| self.catalog.log_dependencies(*id))
+                .map(|id| self.catalog.get_entry(&id).name().item.clone())
+                .collect::<Vec<_>>();
+            if !log_names.is_empty() {
+                return Err(CoordError::UntargetedLogRead { log_names });
+            }
+        }
```

Comment:
There is an edge case where this check can fail do prevent untargeted log peeks. Assume we have a cluster with a single replica and two clients running statements against this cluster.

1. Client A: `SELECT * FROM mz_materializations AS OF <now + 10s>`.
2. Client B: `CREATE CLUSTER REPLICA this_cluster.replica2 ...`.
3. Client A's `SELECT` returns.

Because Client A's `SELECT` was made without `cluster_replica` being set, `replica2` was hydrated with the `Peek` corresponding to that `SELECT`. Which means when the `SELECT` eventually returns, either of the two replicas could have produced the response.

We have two ways to handle this:

1. Accept that we don't handle this edge-case correctly, as it is unlikely to occur anyway.
2. Always set `target_replica` here, if `num_replicas == 1` and we query introspection sources, e.g.:

    ```rust
    let log_names = source_ids
        .iter()
        .flat_map(|id| self.catalog.log_dependencies(*id))
        .map(|id| self.catalog.get_entry(&id).name().item.clone())
        .collect::<Vec<_>>();
    if !log_names.is_empty() {
        if num_replicas == 1 {
            target_replica = compute_instance.replicas_by_id.keys().next().unwrap();
        } else if target_replica.is_none() {
            return Err(CoordError::UntargetedLogRead { log_names });
        }
    }
    ```

I'm not sure if fixing our handling of the above marginal edge-case warrants this additional complexity. Any opinions?

Reply:
If a replica is dropped that still has targeted peeks waiting, nothing happens immediately. The peek is left in the compute controller's list of outstanding peeks and the user's SQL session hangs on the `SELECT`.

Eventually, we should see a `CancelPeek` when the user aborts the `SELECT` or closes their session (the latter isn't the case currently: https://github.com/MaterializeInc/materialize/issues/12546). The `CancelPeek` removes the outstanding peek from the compute controller's state. Then during the next history reduction, the peek is also removed from the command history.